stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1597,52 +1597,294 @@ public extension Tensor where Scalar : Numeric {
+    return .closedRange(
+      Int32(self.lowerBound)...Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension PartialRangeFrom : TensorRangeExpression where Bound == Int {
+  public var tensorRange: TensorRange {
+    return .partialRangeFrom(Int32(self.lowerBound)..., stride: 1)
+  }
+}
+
+extension PartialRangeUpTo : TensorRangeExpression where Bound == Int {
+  public var tensorRange: TensorRange {
+    return .partialRangeUpTo(..<Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension PartialRangeThrough : TensorRangeExpression where Bound == Int {
+  public var tensorRange: TensorRange {
+    return .partialRangeThrough(...Int32(self.upperBound), stride: 1)
+  }
+}
+
+public extension Tensor {
+  @_fixed_layout
+  struct IndexPath {
+    @usableFromInline
+    let begin, end, strides: Tensor<Int32>
+
+    @usableFromInline
+    let beginMask, endMask, ellipsisMask, newAxisMask, squeezeAxisMask: Int64
+
+    @inlinable @inline(__always)
+    public init(
+      begin: Tensor<Int32>, end: Tensor<Int32>, strides: Tensor<Int32>,
+      beginMask: Int64, endMask: Int64, ellipsisMask: Int64, newAxisMask: Int64,
+      squeezeAxisMask: Int64
+    ) {
+      self.begin = begin
+      self.end = end
+      self.strides = strides
+      self.beginMask = beginMask
+      self.endMask = endMask
+      self.ellipsisMask = ellipsisMask
+      self.newAxisMask = newAxisMask
+      self.squeezeAxisMask = squeezeAxisMask
+    }
+  }
+
+  @inlinable @inline(__always)
+  @differentiable(wrt: self, vjp: _vjpSubscript)
+  subscript(_ indexPath: IndexPath) -> Tensor {
+    get {
+      return Raw.stridedSlice(
+        self, begin: indexPath.begin, end: indexPath.end,
+        strides: indexPath.strides, beginMask: indexPath.beginMask,
+        endMask: indexPath.endMask, ellipsisMask: indexPath.ellipsisMask, 
+        newAxisMask: indexPath.newAxisMask,
+        shrinkAxisMask: indexPath.squeezeAxisMask)
+    }
+    set {
+      self = Raw.tensorStridedSliceUpdate(
+        self, begin: indexPath.begin, end: indexPath.end,
+        strides: indexPath.strides, value: newValue,
+        beginMask: indexPath.beginMask, endMask: indexPath.endMask,
+        ellipsisMask: indexPath.ellipsisMask,
+        newAxisMask: indexPath.newAxisMask,
+        shrinkAxisMask: indexPath.squeezeAxisMask)
+    }
+  }
+
+  @inlinable @inline(__always)
+  // TODO: @differentiable(wrt: self)
+  subscript(_ ranges: TensorRangeExpression...) -> Tensor {
+    get {
+      return self[IndexPath(ranges.map { $0.tensorRange })]
+    }
+    set {
+      self[IndexPath(ranges.map { $0.tensorRange })] = newValue
+    }
+  }
+
+  @inlinable @inline(__always)
+  internal func _vjpSubscript(
+    _ indexPath: IndexPath
+  ) -> (Tensor, (Tensor) -> Tensor) {
+    return (self[indexPath], { [shape = shapeTensor] v in
+      Tensor<Scalar>._pullbackSubscript(v, indexPath, shape)
+    })
+  }
+
+  @inlinable @inline(__always)
+  @differentiable(wrt: seed, vjp: _vjpPullbackSubscript)
```

Comment:
Interesting way to make it infinitely differentiable! However, the compiler does not yet support higher order differentiation yet so it won't work. Also, mathematically a pullback should be a linear function, but a method that takes `self` isn't.

The way we plan to support user-defined infinitely differentiable functions is by requiring `@differentiable` attribute on VJP functions, though we haven't figured out the exact inner workings yet.

I think it'd be better if you simplify this to the traditional single-function VJP definition, since all you'll need in the future is likely to mark the VJP function as `@differentiable`. What do you think?

Reply:
I see. Ok, yes we can add this in the future when this is worked out. I wasn't actually sure if this approach would work.